date.go

```diff
@@ -0,0 +1,88 @@
+package imap
+
+import (
+	"fmt"
+	"time"
+)
+
+// Date and time layouts.
+// Dovecot adds a leading zero to dates:
+//   https://github.com/dovecot/core/blob/4fbd5c5e113078e72f29465ccc96d44955ceadc2/src/lib-imap/imap-date.c#L166
+// Cyrus adds a leading space to dates:
+//   https://github.com/cyrusimap/cyrus-imapd/blob/1cb805a3bffbdf829df0964f3b802cdc917e76db/lib/times.c#L543
+const (
+	// Described in RFC 1730 on page 55.
+	DateLayout = "_2-Jan-2006"
+	// Described in RFC 1730 on page 55.
+	DateTimeLayout = "_2-Jan-2006 15:04:05 -0700"
+	// Described in RFC 5322 section 3.3.
+	messageDateTimeLayout = "Mon, 02 Jan 2006 15:04:05 -0700"
```

Comment:
This layout is referred to in RFC 3501 as `env-date` (but is not defined in this RFC, it's just a `nstring`), maybe it's better to name it `envelopeDateTimeLayout`. I think the only place where it's used is in the envelope (if I didn't missed something).